### PR TITLE
Add simple pool for binary protocol ReadBinary

### PR DESCRIPTION
**Pull request by robskillington**

### CODE REVIEW

File: lib/go/thrift/binary_protocol.go [hunk 1]
```diff
@@ -25,8 +25,36 @@ import (
 	"fmt"
 	"io"
 	"math"
+	"sync"
 )
 
+// BytesPoolAlloc is the constant for how big the slices being allocated
+// from the bytes pool are
+const BytesPoolAlloc = 256
```
Comment (prateek) [inline lib/go/thrift/binary_protocol.go]: nit: maybe `MaxBytesPoolAlloc` or something to indicate it's a limit

File: lib/go/thrift/binary_protocol.go [hunk 2]
```diff
@@ -25,8 +25,36 @@ import (
...
+	if cap(b) != BytesPoolAlloc {
+		return false
+	}
+	bytesPool.Put(b[:BytesPoolAlloc])
+	return true
+}
+
+// BytesPoolGet returns a pooled byte slice of capacity BytesPoolAlloc.
+func BytesPoolGet() []byte {
+	return bytesPool.Get().([]byte)[:BytesPoolAlloc]
```
Comment (prateek) [inline lib/go/thrift/binary_protocol.go]: is the `[:BytesPoolAlloc]` necessary here?

**Review by prateek — APPROVED** [review]
LGTM one step closer to our own binary protocol ;)

File: lib/go/thrift/binary_protocol.go (hunk 1 above)
Comment (robskillington) [reply lib/go/thrift/binary_protocol.go]: Sure thing.

File: lib/go/thrift/binary_protocol.go (hunk 2 above)
Comment (robskillington) [reply lib/go/thrift/binary_protocol.go]: Probably not, if we do it on the Put.  I'll remove.

**Pull request merged**